# Branching Programs

The kinds of computations we have been looking at thus far are called straight-line programs. They execute one statement after another in the order in which they appear, and stop when they run out of statements.

Program that can change the order of executing instructions, based on defined *conditional*, is called a *branching program*. A conditional statement has three parts:

- a test, i.e., an expression that evaluates to either `True` or `False`;

- a block of code that is executed if the test evaluates to `True`; and

- an optional block of code that is executed if the test evaluates to `False`.

### Conditional Statement in Python

In Python, a conditional statement has the following form:

```bash
if <condition>:
    <body>
```
Optionally, an optional block code that is executed if the test isn't passed after `else` statement.

```bash
if <condition>:
    <body>
else:
    <body>
```

**Example 1.**

Write a program that prints `x is equal zero` if $x=0$.

In [1]:
x = 0

if x == 0:
    print('x is equal zero')

x is equal zero


In [2]:
x = 1
if x == 0:
    print('x is equal zero')

If $x = 0$, the output will be `x is equal zero`. Otherwise, the program will produce no output. In our super-basic example, this behaviour does not influence our efficiency of programming. But imagine that you present your code to your manager (or lecturer in this case), set $x=0$ and everything is okay. But if $x \neq 0$, your manager won't be able to recognize if something went wrong. So you should handle the case if $x \neq 0$ using `else` statement.

In [3]:
x = 0

if x == 0:
    print('x is equal zero')
else:
    print('x is not equal zero')

x is equal zero


In [4]:
x = 1

if x == 0:
    print('x is equal zero')
else:
    print('x is not equal zero')

x is not equal zero


**Example 2.**

Write a program that for given $x$ will print its absolute value $|x|$.

Remember that absolute value for $x \in \mathbb{R}$ is defined by the following function $f: \mathbb{R} \longmapsto \mathbb{R}$.

$$
f(x) = |x| = \begin{cases}
    x,& \text{ if } x \geq 0 \\
    -x,& \text{ if } x < 0
\end{cases}
$$

For example $|3|=3$, but $|-3| = 3$, so for every $x\in \mathbb{R}$ occurs $|x|=|-x|$ and $|x| \geq 0$.

In [5]:
x = 3

if x >= 0:
    print(x)
else:
    print(-x)

3


In [6]:
x = -3

if x >= 0:
    print(x)
else:
    print(-x)

3


**Example 3.**

Write a program that will print `x is even` if $x$ is even, and `x is odd` otherwise.

There are many ways to solve this problem.

1. Number $x\in \mathbb{Z}$ is even if one of the divisors of $x$ is 2. That implies that $x$ is even if the remainder of $\frac{x}{2}$ equals 0.

In [7]:
x = 4

if x % 2 == 0:
    print('x is even')
else:
    print('x is odd')

x is even


In [8]:
x = 5

if x % 2 == 0:
    print('x is even')
else:
    print('x is odd')

x is odd


2. You can also define the condition in an opposite way. Number $x$ is odd if the remainder $\frac{x}{2}$ does not equal 0.

In [9]:
x = 4

if x % 2 != 0:
    print('x is odd')
else:
    print('x is even')

x is even


In [10]:
x = 5

if x % 2 != 0:
    print('x is odd')
else:
    print('x is even')

x is odd


3. The third way to solve this problem is to use `//` operation. If $x$ is odd, then operation `x / 2` will give a `float` type value. So if $x$ is odd, then `x // 2` will lose its floating part.

In [11]:
x = 4

if (x // 2) * 2 == x:
    print('x is even')
else:
    print('x is odd')

x is even


In [12]:
x = 5

if (x // 2) * 2 == x:
    print('x is even')
else:
    print('x is odd')

x is odd


4. The same in the opposite way.

In [13]:
x = 4

if (x // 2) * 2 != x:
    print('x is odd')
else:
    print('x is even')

x is even


In [14]:
x = 5

if (x // 2) * 2 != x:
    print('x is odd')
else:
    print('x is even')

x is odd


### Flowchart

A **flowchart** is a type of diagram that represents a workflow or process. A flowchart can also be defined as a diagrammatic representation of an algorithm, a step-by-step approach to solving a task.

![Flowcharts](images/flowcharts.png)

**Example 1.**

Draw a flowchart of the program that prints `x is equal zero` if $x=0$.

![Flowchart](images/fc_ex1.png)

**Example 2.**

Draw a flowchart of the program that for given $x$ will print its absolute value $|x|$.

![Flowchart](images/fc_ex2.png)

**Example 3.**

Draw a flowchart of the program that will print `x is even` if $x$ is even, and `x is odd` otherwise.

![Flowchart](images/fc_ex3.png)

### Pseudocode

**Pseudocode** is a universal set of instructions for some procedure that can be understood by every programmer, no matter which programming language they use.

**Example 1.**

Write a pseudocode of the program that prints `x is equal zero` if $x=0$.

```bash
EQUAL_ZERO(int x): str
    if x = 0 then
        print that x equals zero
    else
        print that x not equals zero
```

**Example 2.**

Write a pseudocode of the program that for given $x$ will print its absolute value $|x|$.

```bash
ABSOLUTE_VALUE(int x): int
    if x >= 0 then
        return x
    else
        retuurn -x
```

**Example 3.**

Write a pseudocode of the program that will print `x is even` if $x$ is even, and `x is odd` otherwise.

```bash
IS_EVEN_ODD(int x): str
    if x mod 2 = 0 then
        print that x is even
    else
        print that x is odd
```

### Test Coverage

To be 100% sure that your program works as you expect, check every possible input and every possible output. If somewhere the program gives unexpected results, your program needs repair.

**Example**

Assume you want to write a program that will say which number $x, y, z \in \mathbb{R}$ is the least.

In [15]:
x = 15
y = 5
z = 11

if x < y:
    if x < z:
        print('x is least')
    else:
        print('z is least')
else:
    print('y is least')

y is least


In [16]:
x = 15
y = 13
z = 11

if x < y:
    if x < z:
        print('x is least')
    else:
        print('z is least')
else:
    print('y is least')

y is least


All conditions are correct. For $x = 15$, $y = 5$, $z = 11$ the output is `y is least`. But for $x = 15$, $y = 13$, $z = 11$ the output is also `y is least`. You must pay attention that if the first condition returns `True`, then the `else` block becomes unavailable.

In [17]:
x = 15
y = 5
z = 11

if x < y and x < z:
    print('x is least')
else:
    if y < z:
        print('y is least')
    else:
        print('z is least')

y is least


In [18]:
x = 15
y = 13
z = 11

if x < y and x < z:
    print('x is least')
else:
    if y < z:
        print('y is least')
    else:
        print('z is least')

z is least


For $x = 15$, $y = 5$, $z = 11$ the output is `y is least`, and for $x = 15$, $y = 13$, $z = 11$ the output is also `z is least`.

### Matching

In Python, there exists the structure `elif <condition>` that combines `else: if <condition>` into a shorter form.

**Example**

The program that will say which number $x, y, z \in \mathbb{R}$ is the least can be written more concisely using the `elif` statement.

In [19]:
x = 15
y = 5
z = 11

if x < y and x < z:
    print('x is least')
elif y < z:
    print('y is least')
else:
    print('z is least')

y is least


In [20]:
x = 15
y = 13
z = 11

if x < y and x < z:
    print('x is least')
elif y < z:
    print('y is least')
else:
    print('z is least')

z is least


For $x = 15$, $y = 5$, $z = 11$ the output is `y is least`, and for $x = 15$, $y = 13$, $z = 11$ the output is also `z is least`.

# Iterations

Writing a program that asks the user how many time he wants to print the letter X, and then prints a string with that number of X’s. We might think about writing something like this.

```bash
numXs = int(input('How many times should I print the letter X? ')) 
toPrint = None
if numXs == 1:
    toPrint = 'X'
elif numXs == 2:
    toPrint = 'XX'
elif numXs == 3:
    toPrint = 'XXX'
#...
print(toPrint)
```

But it would quickly become apparent that we would need as many conditionals as there are positive integers and there are an infinite number of those. 

### Loop

When we want a program to do the same thing many times, we can use **iteration**. Like a conditional statement, it begins with a test. If the test evaluates to `True`, the program executes the **loop** body once, and then goes back to reevaluate the test. This process is repeated until the test evaluates to `False`, after which control passes to the code following the iteration statement.

In Python, a iteration has the following form.

```bash
while <condition>:
    <body>
```

**Example 1.**

Write a program that prints numbers from 1 to $n - 1$.

In [21]:
i = 1
n = 10

while i < n:
    print(i)
    i = i + 1

1
2
3
4
5
6
7
8
9


In [22]:
i = 1
n = 10

condition = i < n

while condition:
    print(i)
    i = i + 1
    condition = i < n

1
2
3
4
5
6
7
8
9


There are two methods of evaluating loop condition. It is necessary to be able use appropriate method due to program application. In this case using first or second method does not matter.

```bash
FROM_ONE_TO_NINE(int n): str
    i <- 1
    while i < n do
        print i
        i <- i + 1
```

![Flowchart](images/one_to_n.png)

**Example 2.**

Write a program that prints even numbers from 1 to $n - 1$.

In [23]:
i = 1
n = 10

while i < n:
    if i % 2 == 0:
        print(i)
    i = i + 1

2
4
6
8


```bash
FROM_ONE_TO_NINE_EVEN(int n): str
    i <- 1
    while i < n do
        if i mod 2 = 0 then
            print i
        i <- i + 1
```

![Flowchart](images/one_to_n_even.png)

### Iterative Program Rules

1. Choose a variable that *counts*.

2. Initialize the **counter** outside the *loop*.

3. Set up the **end test**.

4. Change the value of the counter inside the *loop*.

**Example 1.**

Use Python to calculate the value of $\displaystyle\sum_{i=1}^{n}{i} = 1 + 2 + 3 + ... + n$.

In [24]:
i = 1
n = 10
ans = 0

while i <= n:
    ans = ans + i
    i = i + 1

print(ans)

55


For **addition**, basic value of the counter is $0$.

**Example 2.**

Use Python to calculate the value of factorial $n!= \displaystyle\prod_{i=1}^{n}{i} = 1 \times 2 \times 3 \times ... \times n$.

In [25]:
i = 1
n = 10
ans = 1

while i <= n:
    ans = ans * i
    i = i + 1

print(ans)

3628800


For **multiplication**, basic value of the counter is $1$.

**Example 3.**

Use Python to calculate the value of $\displaystyle\sum_{i=1}^{n}{i^2} = 1^2 + 2^2 + 3^2 + ... + n^2$.

In [26]:
i = 1
n = 10
ans = 0

while i <= n:
    ans = ans + (i ** 2)
    i = i + 1
    
print(ans)

385


You can also replace `ans = ans + (i ** 2)` with `ans = ans + (i * i)`.

# Simulating Iterations

Sometimes, despite great efforts, it is difficult to find an error by looking at successive lines of written code. In the case of larger projects, this is incredibly troublesome. Therefore, a good programmer must be able to simulate all the algorithms they have written on a piece of paper. Moreover, this is the fastest way to find potential errors in the program.

**Example 1.**

Consider the program that calculates $x^y$ for given $x, y\in \N^{+}$.

In [27]:
x = 3
y = 5
ans = 1

while y > 0:
    ans = ans * x
    y = y - 1
    
print('x^y =', ans)

x^y = 243


Let's simulate, how it works for $x=3$ and $y=5$.

$$\begin{array}{c|c|c|c|c}
    \text{iteration} & x & y & \text{ans} & y > 0 \\
    \hline
    0 & 3 & 5 & 1 & \texttt{True} \\
    1 & 3 & 4 & 3 & \texttt{True} \\
    2 & 3 & 3 & 9 & \texttt{True} \\
    3 & 3 & 2 & 27 & \texttt{True} \\
    4 & 3 & 1 & 81& \texttt{True} \\
    5 & 3 & 0 & 243 & \texttt{False} \\
    \end{array}$$

The answer is $243$.

**Example 2.**

Consider the following flowchart.

![Flowchart 1](images/sim1.png)

The idea is to check whether $n$ is a perfect square, that means $\sqrt{n}\in \mathbb{N}$. For example, $16$ is a perfect square because $\sqrt{16}=4$ and $4 \in \mathbb{N}$. But $20$ is not a perfect square because $\sqrt{20}=2\sqrt{5}$ and $2\sqrt{5} \notin \mathbb{N}$. Let's try it for $n=16$.

$$
\begin{array}{c|c|c|c|c}
    \text{iteration} & n & \text{ans} & \text{ans}^2 & \text{ans}^2 \leq n \\
    \hline
    0 & 16 & 0 & 0 & \texttt{True} \\
    1 & 16 & 1 & 1 & \texttt{True} \\
    2 & 16 & 2 & 4 & \texttt{True} \\
    3 & 16 & 3 & 9 & \texttt{True} \\
    4 & 16 & 4 & 16 & \texttt{True} \\
    5 & 16 & 5 & 25 & \texttt{False} 
\end{array}
$$

So the result is $5$ which we didn't expect. The problem is with the conditional $\text{ans}^2 \leq n$. If we change this to $\text{ans}^2 < n$, the algorithm should work correctly.

![Flowchart 2](images/sim2.png)

$$
\begin{array}{c|c|c|c|c}
    \text{iteration} & n & \text{ans} & \text{ans}^2 & \text{ans}^2 < n \\
    \hline
    0 & 16 & 0 & 0 & \texttt{True} \\
    1 & 16 & 1 & 1 & \texttt{True} \\
    2 & 16 & 2 & 4 & \texttt{True} \\
    3 & 16 & 3 & 9 & \texttt{True} \\
    4 & 16 & 4 & 16 & \texttt{False} 
\end{array}
$$

So the result is $4$ as we expected, but how would the algorithm behave for $n=20$?

$$
\begin{array}{c|c|c|c|c}
    \text{iteration} & n & \text{ans} & \text{ans}^2 & \text{ans}^2 < n \\
    \hline
    0 & 20 & 0 & 0 & \texttt{True} \\
    1 & 20 & 1 & 1 & \texttt{True} \\
    2 & 20 & 2 & 4 & \texttt{True} \\
    3 & 20 & 3 & 9 & \texttt{True} \\
    4 & 20 & 4 & 16 & \texttt{True} \\
    5 & 20 & 5 & 25 & \texttt{False} 
\end{array}
$$

The result is $5$ which is not correct, because $20$ is not a perfect square. Our goal is to handle the case if the answer (ans) is bigger than expected.

![Flowchart 3](images/sim3.png)

$$
\begin{array}{c|c|c|c|c|c}
    \text{iteration} & n & \text{ans} & \text{ans}^2 & \text{ans}^2 < n & \text{ans}^2 = n \\
    \hline
    0 & 16 & 0 & 0 & \texttt{True} & \texttt{False} \\
    1 & 16 & 1 & 1 & \texttt{True} & \texttt{False} \\
    2 & 16 & 2 & 4 & \texttt{True} & \texttt{False} \\
    3 & 16 & 3 & 9 & \texttt{True} & \texttt{False} \\
    4 & 16 & 4 & 16 & \texttt{False} & \texttt{True} 
\end{array}
$$

For $n=16$, output says that it is `True` and is for $4$.

$$
\begin{array}{c|c|c|c|c|c}
    \text{iteration} & n & \text{ans} & \text{ans}^2 & \text{ans}^2 < n & \text{ans}^2 = n \\
    \hline
    0 & 20 & 0 & 0 & \texttt{True} & \texttt{False} \\
    1 & 20 & 1 & 1 & \texttt{True} & \texttt{False} \\
    2 & 20 & 2 & 4 & \texttt{True} & \texttt{False} \\
    3 & 20 & 3 & 9 & \texttt{True} & \texttt{False} \\
    4 & 20 & 4 & 16 & \texttt{True} & \texttt{False} \\
    5 & 20 & 5 & 25 & \texttt{False} & \texttt{False} 
\end{array}
$$

For $n=20$, output says that it is `False`.

# Optimizing Amount of Calculations

**Example 1.**

We want to check how many numbers are the divisors of $n$. Consider the following code.

In [28]:
n = 30
i = 1
ans = 0

while i <= n:
    if n % i == 0:
        ans = ans + 1
    i = i + 1

print(ans)

8


Program checks every number from 1 to $n$ and it works, but imagine how many operations this program could do if you insert the very big number. From mathematics we know that the largest divisor of $n$ is itself, and the second largest divisor cannot be bigger than $\frac{n}{2}$. To reduce amount of calculations we should change our condition.

In [29]:
n = 30
i = 1
ans = 0

while i <= n / 2:
    if n % i == 0:
        ans = ans + 1
    i = i + 1

ans = ans + 1

print(ans)

8


Now this program returns the same outputs, but the amount of operations is reduced by a half.

**Example 2.**

We want to check whether $n$ is a prime number. Consider the following pseudocode.

```bash
IS_PRIME(int n): bool
    i <- 2
    ans <- true
    while i < n do
        if n mod i = 0 then
            ans <- False
        i <- i + 1
    return ans
```

There occurs the same problem. There are too many calculations, because from mathematics we know that the largest divisor of $n$ is itself, and the second largest divisor cannot be bigger than $\frac{n}{2}$.

```bash
IS_PRIME(int n): bool
    i <- 2
    ans <- true
    while i <= n / 2 do
        if n mod i = 0 then
            ans <- False
        i <- i + 1
    return ans
```

We reduced amount of calculations by a half, and we can still achieve better results. If $i$ is even we know that 2 is its divisor so we can skip every even number.

```bash
IS_PRIME(int n): bool
    i <- 2
    ans <- true
    while i <= n / 2 do
        if n mod i = 0 then
            ans <- False
        if i = 2 then
            i <- i + 1
        else
            i <- i + 2
    return ans
```

Now our program works very fast, don't do any unnecessary operations and skips numbers, that can longer time of calculations. But there is still one unfixed problem. Imagine that we enter some previous number like 4, then program will calculate all next possible numbers instead of returning `False`.
```bash
IS_PRIME(int n): bool
    i <- 2
    ans <- true
    while ans = true and i <= n / 2 do
        if n mod i = 0 then
            ans <- False
        if i = 2 then
            i <- i + 1
        else
            i <- i + 2
    return ans
```

This program is the most efficient way of calculating whether $n\in \mathbb{N}$ is a prime number.

# Halting Problem

In computability theory, the \textit{halting problem} is the problem of determining, from a description of an arbitrary computer program and an input, whether the program will finish running, or continue to run forever. The halting problem is undecidable, meaning that no general algorithm exists that solves the halting problem for all possible program–input pairs.

**Example 1.**

Consider the following code.

In [30]:
n = 30
i = 1

while i <= n:
    if i % 3 == 0:
        print('Fizz')
    elif i % 5 == 0:
        print('Buzz')
    elif i % 3 == 0 and i % 3 == 0:
        print('FizzBuzz')

KeyboardInterrupt: 

Looks completely normal but if you run, its output will never appear so there is the halting problem. That's because `i = i + 1` statement is missing. Moreover, there is a semantic error that makes the last condition unavailable.

**Example 2.**

Consider the following code.

In [31]:
n = 30
i = 1

while i <= n:
    if n % i == 0:
        print(i, 'is a divisor of', n)
    i = i + 1

1 is a divisor of 30
2 is a divisor of 30
3 is a divisor of 30
5 is a divisor of 30
6 is a divisor of 30
10 is a divisor of 30
15 is a divisor of 30
30 is a divisor of 30


This program does not have the halting problem. Looks sematically correct, but it is not. If $n < 1$, then program will produce no output.

In [32]:
n = -30
i = 1

while i <= n:
    if n % i == 0:
        print(i, 'is a divisor of', n)
    i = i + 1